vms/platformvm/lock/camino_lock.go

```diff
@@ -0,0 +1,151 @@
+// Copyright (C) 2022, Chain4Travel AG. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package lock
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/vms/components/avax"
+)
+
+var (
+	errInvalidLockState = errors.New("invalid lockState")
+	errNestedLocks      = errors.New("shouldn't nest locks")
+	ThisTxID            = ids.ID{'t', 'h', 'i', 's', ' ', 't', 'x', ' ', 'i', 'd'}
+)
+
+type LockState byte
+
+const (
+	LockStateUnlocked        LockState = 0b00
+	LockStateDeposited       LockState = 0b01
+	LockStateBonded          LockState = 0b10
+	LockStateDepositedBonded LockState = 0b11
+)
+
+var lockStateStrings = map[LockState]string{
+	LockStateUnlocked:        "unlocked",
+	LockStateDeposited:       "deposited",
+	LockStateBonded:          "bonded",
+	LockStateDepositedBonded: "depositedBonded",
+}
+
+func (ls LockState) String() string {
+	lockStateString, ok := lockStateStrings[ls]
+	if !ok {
+		return fmt.Sprintf("unknownLockState(%d)", ls)
+	}
+	return lockStateString
+}
+
+func (ls LockState) Verify() error {
+	if ls < LockStateUnlocked || LockStateDepositedBonded < ls {
+		return errInvalidLockState
+	}
+	return nil
+}
+
+func (ls LockState) IsBonded() bool {
+	return LockStateBonded&ls == LockStateBonded
+}
+
+func (ls LockState) IsDeposited() bool {
+	return LockStateDeposited&ls == LockStateDeposited
+}
+
+type LockIDs struct {
+	DepositTxID ids.ID `serialize:"true" json:"depositTxID"`
+	BondTxID    ids.ID `serialize:"true" json:"bondTxID"`
+}
+
+func (lock LockIDs) LockState() LockState {
+	lockState := LockStateUnlocked
+	if lock.DepositTxID != ids.Empty {
+		lockState = LockStateDeposited
+	}
+	if lock.BondTxID != ids.Empty {
+		lockState |= LockStateBonded
+	}
+	return lockState
+}
+
+func (lock LockIDs) Lock(lockState LockState) LockIDs {
+	if lockState.IsDeposited() {
+		lock.DepositTxID = ThisTxID
+	}
```

Comment:
return / else ??

Reply:
intentionally, so it will set full state (d&b) - we restrict state to "half-state" in places where we use it, making sure that applied/removed lockState is only deposited or bonded

Reply:
well, cause there probably won't be any use for full-state and the whole state concept looks a bit obsolete now, I think its better make it with return/else to not let any assumptions, yep

Reply:
Could be used for double locking (1TX == deposit + bond)